README
This is the repository for the CPAN Marpa::PP distribution.  The
distribution directory is the pp/ directory.


pp/Changes
# Copyright 2012 Jeffrey Kegler
# This file is part of Marpa::PP.  Marpa::PP is free software: you can
# redistribute it and/or modify it under the terms of the GNU Lesser
# General Public License as published by the Free Software Foundation,
# either version 3 of the License, or (at your option) any later version.
#
# Marpa::PP is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser
# General Public License along with Marpa::PP.  If not, see
# http://www.gnu.org/licenses/.

Revision history for Marpa::PP

0.014000 Mon Mar 12 09:22:04 PDT 2012

  * Fix various installation issues.

  * PP distribution now in separate git repository

0.012000 Tue Nov 22 00:20:33 PST 2011

    * Eliminated deprecated interfaces from tests

    * Eliminated use of bare Marpa:: namespace
      in favor of Marpa::PP::

0.010000 Tue Oct 25 08:56:47 PDT 2011

     * Tracking Marpa::XS

     * Differs from Marpa-XS-0.018000: no ranking methods

0.008000
     First official release (ALPHA).


pp/lib/Marpa/PP/Callback.pm
# Copyright 2012 Jeffrey Kegler
# This file is part of Marpa::PP.  Marpa::PP is free software: you can
# redistribute it and/or modify it under the terms of the GNU Lesser
# General Public License as published by the Free Software Foundation,
# either version 3 of the License, or (at your option) any later version.
#
# Marpa::PP is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser
# General Public License along with Marpa::PP.  If not, see
# http://www.gnu.org/licenses/.

package Marpa::PP::Callback;

use 5.010;
use warnings;
use strict;
use integer;

use vars qw($VERSION $STRING_VERSION);
$VERSION        = '0.014000';
$STRING_VERSION = $VERSION;
{
## no critic (BuiltinFunctions::ProhibitStringyEval)
## no critic (ValuesAndExpressions::RequireConstantVersion)
    $VERSION = eval $VERSION;
}

package Marpa::PP::Internal::Callback;

use English qw( -no_match_vars );

sub Marpa::PP::location {
    Marpa::PP::exception('No context for location callback')
        if not my $context = $Marpa::PP::Internal::CONTEXT;
    my ( $context_type, $and_node, $recce ) = @{$context};
    if ( $context_type eq 'and-node' ) {
        my $earleme =
            $and_node->[Marpa::PP::Internal::And_Node::START_EARLEME];
        my $earley_sets =
            $recce->[Marpa::PP::Internal::Recognizer::EARLEY_SETS];
        return $earley_sets->[$earleme]
            ->[Marpa::PP::Internal::Earley_Set::ORDINAL];
    } ## end if ( $context_type eq 'and-node' )
    Marpa::PP::exception('LOCATION called outside and-node context');
} ## end sub Marpa::PP::location

sub Marpa::PP::cause_location {
    Marpa::PP::exception('No context for cause_location callback')
        if not my $context = $Marpa::PP::Internal::CONTEXT;
    my ( $context_type, $and_node, $recce ) = @{$context};
    if ( $context_type eq 'and-node' ) {
        my $earleme =
            $and_node->[Marpa::PP::Internal::And_Node::CAUSE_EARLEME];
        my $earley_sets =
            $recce->[Marpa::PP::Internal::Recognizer::EARLEY_SETS];
        return $earley_sets->[$earleme]
            ->[Marpa::PP::Internal::Earley_Set::ORDINAL];
    } ## end if ( $context_type eq 'and-node' )
    Marpa::PP::exception('cause_location() called outside and-node context');
} ## end sub Marpa::PP::cause_location

no strict 'refs';
*{'Marpa::PP::token_location'} = \&Marpa::PP::cause_location;
use strict;

1;


pp/lib/Marpa/PP/Internal.pm
# Copyright 2012 Jeffrey Kegler
# This file is part of Marpa::PP.  Marpa::PP is free software: you can
# redistribute it and/or modify it under the terms of the GNU Lesser
# General Public License as published by the Free Software Foundation,
# either version 3 of the License, or (at your option) any later version.
#
# Marpa::PP is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser
# General Public License along with Marpa::PP.  If not, see
# http://www.gnu.org/licenses/.

package Marpa::PP::Internal;

use 5.010;
use strict;
use warnings;
use integer;
use Carp;

use vars qw($VERSION $STRING_VERSION);
$VERSION        = '0.014000';
$STRING_VERSION = $VERSION;
{
## no critic (BuiltinFunctions::ProhibitStringyEval)
## no critic (ValuesAndExpressions::RequireConstantVersion)
    $VERSION = eval $VERSION;
}

*Marpa::PP::exception = \&Carp::croak;

sub Marpa::PP::internal_error {
    Carp::confess(
        "Internal Marpa::PP Error: This could be a bug in Marpa::PP\n", @_ );
}

# Perl critic at present is not smart about underscores
# in hex numbers
## no critic (ValuesAndExpressions::RequireNumberSeparators)
use constant N_FORMAT_MASK     => 0xffff_ffff;
use constant N_FORMAT_HIGH_BIT => 0x8000_0000;
## use critic

# Also used as mask, so must be 2**n-1
# Perl critic at present is not smart about underscores
# in hex numbers
use constant N_FORMAT_MAX => 0x7fff_ffff;

sub Marpa::PP::offset {
    my (@desc) = @_;
    my @fields = ();
    for my $desc (@desc) { push @fields, split q{ }, $desc; }
    my $pkg        = caller;
    my $prefix     = $pkg . q{::};
    my $offset     = -1;
    my $in_comment = 0;

    no strict 'refs';
    FIELD: for my $field (@fields) {

        if ($in_comment) {
            $in_comment = $field ne ':}' && $field ne '}';
            next FIELD;
        }

        PROCESS_OPTION: {
            last PROCESS_OPTION if $field !~ /\A [{:] /xms;
            if ( $field =~ / \A [:] package [=] (.*) /xms ) {
                $prefix = $1 . q{::};
                next FIELD;
            }
            if ( $field =~ / \A [:]? [{] /xms ) {
                $in_comment++;
                next FIELD;
            }
        } ## end PROCESS_OPTION:

        if ( $field !~ s/\A=//xms ) {
            $offset++;
        }

        if ( $field =~ / \A ( [^=]* ) = ( [0-9+-]* ) \z/xms ) {
            $field  = $1;
            $offset = $2 + 0;
        }

        Marpa::PP::exception("Unacceptable field name: $field")
            if $field =~ /[^A-Z0-9_]/xms;
        my $field_name = $prefix . $field;
        *{$field_name} = sub () {$offset};
    } ## end for my $field (@fields)
    return 1;
} ## end sub Marpa::PP::offset

1;


pp/lib/Marpa/PP.pm
# Copyright 2012 Jeffrey Kegler
# This file is part of Marpa::PP.  Marpa::PP is free software: you can
# redistribute it and/or modify it under the terms of the GNU Lesser
# General Public License as published by the Free Software Foundation,
# either version 3 of the License, or (at your option) any later version.
#
# Marpa::PP is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser
# General Public License along with Marpa::PP.  If not, see
# http://www.gnu.org/licenses/.

package Marpa::PP;

use 5.010;
use strict;
use warnings;

use vars qw($VERSION $STRING_VERSION $DEBUG);
$VERSION        = '0.014000';
$STRING_VERSION = $VERSION;
{
## no critic (BuiltinFunctions::ProhibitStringyEval)
## no critic (ValuesAndExpressions::RequireConstantVersion)
    $VERSION = eval $VERSION;
}
$DEBUG = 0;

use Carp;
use English qw( -no_match_vars );

# Die if more than one of the Marpa modules is loaded
BEGIN {
    if ( defined $Marpa::VERSION ) {
        Carp::croak( "You can only load one of the Marpa modules at a time\n",
            'Version ', $Marpa::VERSION, " of Marpa is already loaded\n" );
    }
    if ( defined $Marpa::XS::VERSION ) {
        Carp::croak( "You can only load one of the Marpa modules at a time\n",
            'Version ', $Marpa::XS::VERSION,
            " of Marpa::XS is already loaded\n" );
    }
} ## end BEGIN

use Marpa::PP::Version;

@Marpa::PP::CARP_NOT = ();
for my $middle ( q{}, '::Internal' ) {
    for my $end ( q{}, qw(::Recognizer ::Callback ::Grammar ::Value) ) {
        push @Marpa::PP::CARP_NOT, 'Marpa::PP' . $middle . $end;
    }
}
PACKAGE: for my $package (@Marpa::PP::CARP_NOT) {
    no strict 'refs';
    next PACKAGE if $package eq 'Marpa';
    *{ $package . q{::CARP_NOT} } = \@Marpa::PP::CARP_NOT;
}

if ( not $ENV{'MARPA_AUTHOR_TEST'} ) {
    $Marpa::PP::DEBUG = 0;
}
else {
    $Marpa::PP::DEBUG = 1;
}

sub version_ok {
    my ($sub_module_version) = @_;
    return 'not defined' if not defined $sub_module_version;
    return "$sub_module_version does not match Marpa::PP::VERSION " . $VERSION
        if $sub_module_version != $VERSION;
    return;
} ## end sub version_ok

my $version_result;
require Marpa::PP::Internal;
( $version_result = version_ok($Marpa::PP::Internal::VERSION) )
    and die 'Marpa::PP::Internal::VERSION ', $version_result;

require Marpa::PP::Grammar;
( $version_result = version_ok($Marpa::PP::Grammar::VERSION) )
    and die 'Marpa::PP::Grammar::VERSION ', $version_result;

require Marpa::PP::Recognizer;
( $version_result = version_ok($Marpa::PP::Recognizer::VERSION) )
    and die 'Marpa::PP::Recognizer::VERSION ', $version_result;

require Marpa::PP::Value;
( $version_result = version_ok($Marpa::PP::Value::VERSION) )
    and die 'Marpa::PP::Value::VERSION ', $version_result;

require Marpa::PP::Callback;
( $version_result = version_ok($Marpa::PP::Callback::VERSION) )
    and die 'Marpa::PP::Callback::VERSION ', $version_result;

$Marpa::USING_PP = 1;
$Marpa::USING_XS = undef;
$Marpa::MODULE   = 'Marpa::PP';

*Marpa::Grammar::check_terminal = \&Marpa::PP::Grammar::check_terminal;
*Marpa::Grammar::new            = \&Marpa::PP::Grammar::new;
*Marpa::Grammar::precompute     = \&Marpa::PP::Grammar::precompute;
*Marpa::Grammar::set            = \&Marpa::PP::Grammar::set;
*Marpa::Grammar::show_AHFA      = \&Marpa::PP::Grammar::show_AHFA;
*Marpa::Grammar::show_NFA       = \&Marpa::PP::Grammar::show_NFA;
*Marpa::Grammar::show_accessible_symbols =
    \&Marpa::PP::Grammar::show_accessible_symbols;
*Marpa::Grammar::show_nullable_symbols =
    \&Marpa::PP::Grammar::show_nullable_symbols;
*Marpa::Grammar::show_nulling_symbols =
    \&Marpa::PP::Grammar::show_nulling_symbols;
*Marpa::Grammar::show_productive_symbols =
    \&Marpa::PP::Grammar::show_productive_symbols;
*Marpa::Grammar::show_problems     = \&Marpa::PP::Grammar::show_problems;
*Marpa::Grammar::show_rules        = \&Marpa::PP::Grammar::show_rules;
*Marpa::Grammar::show_symbols      = \&Marpa::PP::Grammar::show_symbols;
*Marpa::Recognizer::alternative    = \&Marpa::PP::Recognizer::alternative;
*Marpa::Recognizer::check_terminal = \&Marpa::PP::Recognizer::check_terminal;
*Marpa::Recognizer::current_earleme =
    \&Marpa::PP::Recognizer::current_earleme;
*Marpa::Recognizer::earleme_complete =
    \&Marpa::PP::Recognizer::earleme_complete;
*Marpa::Recognizer::earley_set_size =
    \&Marpa::PP::Recognizer::earley_set_size;
*Marpa::Recognizer::end_input = \&Marpa::PP::Recognizer::end_input;
*Marpa::Recognizer::exhausted = \&Marpa::PP::Recognizer::exhausted;
*Marpa::Recognizer::latest_earley_set =
    \&Marpa::PP::Recognizer::latest_earley_set;
*Marpa::Recognizer::new         = \&Marpa::PP::Recognizer::new;
*Marpa::Recognizer::parse_count = \&Marpa::PP::Recognizer::parse_count;
*Marpa::Recognizer::read        = \&Marpa::PP::Recognizer::read;
*Marpa::Recognizer::reset_evaluation =
    \&Marpa::PP::Recognizer::reset_evaluation;
*Marpa::Recognizer::set            = \&Marpa::PP::Recognizer::set;
*Marpa::Recognizer::show_and_nodes = \&Marpa::PP::Recognizer::show_and_nodes;
*Marpa::Recognizer::show_earley_sets =
    \&Marpa::PP::Recognizer::show_earley_sets;
*Marpa::Recognizer::show_or_nodes = \&Marpa::PP::Recognizer::show_or_nodes;
*Marpa::Recognizer::show_progress = \&Marpa::PP::Recognizer::show_progress;
*Marpa::Recognizer::status        = \&Marpa::PP::Recognizer::status;
*Marpa::Recognizer::terminals_expected =
    \&Marpa::PP::Recognizer::terminals_expected;
*Marpa::Recognizer::tokens = \&Marpa::PP::Recognizer::tokens;
*Marpa::Recognizer::value  = \&Marpa::PP::Recognizer::value;
*Marpa::location           = \&Marpa::PP::location;
*Marpa::token_location     = \&Marpa::PP::token_location;

1;

__END__
